### ISSUE TITLE
"Unassigned local variable" is reported for a variable, which is always assigned

### ISSUE THREAD

**Issue by DoctorKrolic** [dotnet/roslyn#80172]
**Version Used**:
VS 17.14.13

**Steps to Reproduce**:
```cs
int val;

try
{
	val = 1;
}
catch
{
	val = 2;
}
finally
{
    Console.WriteLine(val); // Use of unassigned local variable 'val'
}
```

**Diagnostic Id**:
`CS0165`

**Expected Behavior**:
No error since `val` is assigned in both `try` and `catch` "branches", so it is 100% assigned at the point of `finally`

**Actual Behavior**:
Compiler error `CS0165: Use of unassigned local variable 'val'`

**Comment by CyrusNajmabadi** [comment]
I'm not sure if this is incorrect.  The compiler doesn't know if an errant thread .abort will happen prior to either/both val assignments.

The finally truly could run without the assignment happening.

basically, the end of the try/catch assignments are not guaranteed to be reachable, and as such the val may not be assigned.  But the finally will still run as execution exits due to that abort.

**Comment by CyrusNajmabadi** [comment]
@jaredpar  for assessment.

**Comment by RikkiGibson** [comment]
The assignments in 'try' and 'catch' do not necessarily flow to the 'finally' block. I believe there is an assumption that the point at the start of the 'try' or the 'catch' can branch directly to the 'finally'. You could imagine a `Thread.Abort()` call causing this to occur.

https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/variables.md#94415-try-finally-statements

> For a statement *stmt* of the form:
> 
> ```csharp
> try «try_block» finally «finally_block»
> ```
> 
> ...
> - The definite-assignment state of *v* at the beginning of *finally_block* is the same as the definite-assignment state of *v* at the beginning of *stmt*.

And the following section states that `try-catch-finally` is analyzed as if it were `try { try-catch; } finally { }`.

**Comment by RikkiGibson** [comment]
Closing as by design.
